// @stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/add_package_to_scope.js
/**
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// VARIABLES //

// Regular expression to match function declarations such as "declare function abs( x: number ): number;" (captures function name):
var RE_DECLARE_FUNCTION = /declare\s+function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*[<(]/;

// Regular expression to match variable declarations such as "declare var someVar: SomeType;" (captures variable name):
var RE_DECLARE_VAR = /declare\s+var\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:/;

// Regular expression to match class declarations such as "declare class Complex64Array {" (captures class name):
var RE_DECLARE_CLASS = /declare\s+class\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s/;

// Regular expression to match const declarations such as "declare const PI: number;" (captures constant name):
var RE_DECLARE_CONST = /declare\s+const\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:/;

// Regular expression to match variable declarations with interface types such as "declare var ctor: Int32Vector;" (captures variable name and interface name):
var RE_DECLARE_VAR_INTERFACE = /declare\s+(?:var|const)\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:\s*([A-Z][a-zA-Z0-9_$]*)/;


// MAIN //

/**
* Adds a package export to the scope based on TypeScript declarations.
*
* @private
* @param {Object} scope - VM scope object to add the package export to
* @param {*} pkg - package export value to be added to scope
* @param {string} sourceText - TypeScript declaration source text to parse for identifier names
*/
function addPackageToScope( scope, pkg, sourceText ) {
	var interfaceMatch;
	var namespaceMatch;
	var pkgType;
	var match;

	pkgType = typeof pkg;
	if ( pkgType === 'function' ) {
		match = sourceText.match( RE_DECLARE_FUNCTION ) || sourceText.match( RE_DECLARE_VAR ) || sourceText.match( RE_DECLARE_CLASS ); // eslint-disable-line max-len
		if ( match ) {
			scope[ match[1] ] = pkg;
		}
		interfaceMatch = sourceText.match( RE_DECLARE_VAR_INTERFACE );
		if ( interfaceMatch ) {
			// Make the function available under both the variable and interface names:
			scope[ interfaceMatch[1] ] = pkg; // e.g., ctor
			scope[ interfaceMatch[2] ] = pkg; // e.g., Int32Vector
		}
	} else {
		// For objects, check if declared with interface pattern (e.g., `declare var ns: Namespace;`)...
		if ( pkgType === 'object' && pkg !== null ) {
			namespaceMatch = sourceText.match( RE_DECLARE_VAR_INTERFACE );
			if ( namespaceMatch ) {
				scope[ namespaceMatch[1] ] = pkg;
			}
			// Note: intentional fall-through to handle multiple patterns for objects...
		}
		// For all non-function types (including objects), also check const pattern (e.g., `declare const PI: number;` or `declare const obj: {...};`)...
		match = sourceText.match( RE_DECLARE_CONST );
		if ( match ) {
			scope[ match[1] ] = pkg;
		}
	}
}


// EXPORTS //

module.exports = addPackageToScope;
